proposals/269/transport.md

```diff
@@ -0,0 +1,92 @@
+---
+Status: IN PROGRESS  
+Author: Julien Le Dem (original issue) / Maciej Obuchowski (this document)  
+Created: 2021-10-18  
+Issue: [269](https://github.com/OpenLineage/OpenLineage/issues/269)
+---
+
+##Purpose
+
+The goal is to facilitate sending `OpenLineage` events over various protocols. 
+(HTTP, queues, ...) independently of the logic via exposing `Transport` interface.
+
+## Proposed interface
+
+In pseudocode:
+```
+interface Transport {
```

Comment:
It's really nice when, in case of misconfiguration, a system notifies the issue immediately after the start not waiting for `send` method to be called. Perhaps some sort of `init` or `verify` method would help with this.

Imagine you provide Kafka producer configuration with broker urls, encoder classes, SSL stuff to authenticate, etc. You really would like to see if the configuration is correct before job executes and real event is emitted.

Reply:
I agree, this should be done on init, but I think `TransportFactory` described below is a better place for that.

Reply:
Though not ideal, it's common to error on first request. Take for example a simple http client. Usually, errors will be thrown on the first use of the client. This can be for a number of reasons: the host or port aren't valid, typos, etc. That said, I don't see anything wrong with adding a step to verify that the transport properties are valid and the backend is reachable, it just comes at a cost of maintaining such code. I'd feel more natural to know if something is wrong with the transport on first use.

Reply:
I think our options are

1) don't have special handling of first request
2) optionally, make "special" request on init unique to each transport to find if backend is reachable and configuration (like apiKey) is okay
3) have some special handling on first request

I don't think 3) makes a lot of sense. 2) depends very much on Transport, so I think it's best left out for particular implementations. I don't think we can do this in general case, for example for HTTP.